---
title: "Encoding Genetic Data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Encoding Genetic Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(STAVE)
if (requireNamespace("kableExtra", quietly = TRUE)) {
    library(kableExtra)
}
if (requireNamespace("dplyr", quietly = TRUE)) {
    library(dplyr)
}
```

One of the more nuanced aspects of using STAVE is learning how to correctly format genetic data in the *Counts* table. STAVE is designed with the flexibility to encode full haplotypes, preserving information about linkage across multiple genes and multiple loci. But what happens when not all loci amplify in some samples? Or when published data only report locus-by-locus results, rather than full haplotypes? These situations are entirely manageable once we understand the underlying data structure.

### Encoding single-locus data

Imagine you come across a paper detailing *pfcrt* mutations at codons 72 to 76 for a sample of 100 individuals. The CVIET haplotype, commonly observed at these positions, is well-established as a marker of chloroquine resistance. However, rather than presenting haplotype-level data, the paper provides only site-specific information: the number of individuals tested at each codon and the frequency of each observed variant:

```{r, echo=FALSE}
# read in data from inst/extdata
file_path <- system.file("extdata", "vignettes/example_01.xlsx", package = "STAVE")
input_studies <- readxl::read_excel(file_path, sheet = "studies")
input_surveys <- readxl::read_excel(file_path, sheet = "surveys")
input_counts <- readxl::read_excel(file_path, sheet = "counts")

input_counts |>
  filter(study_key == "study_01") |>
  kbl(format = "html", table.attr = "style='width:100%; white-space: nowrap;'") |>
  kable_styling(bootstrap_options = c("hover", "condensed"), 
                full_width = TRUE)
```

Notice that the `total_num` is not consistent over loci. This is very common, and is usually caused by some loci failing to amplify in some samples.

This method of encoding genetic data is straightforward and easy to interpret. For instance, the prevalence of glutamic acid (“E”) at codon 75 can be directly calculated as 20 out of 92 samples, or approximately 22%. However, this approach sacrifices information about full haplotypes. Specifically, we cannot determine whether the 20 samples with `crt:75:E` are the same individuals who also carry `crt:76:T`. As a result, we are limited to estimating single-locus prevalences and cannot infer multi-locus patterns or linkage. While this encoding is valid, it is suboptimal for analyses that rely on haplotype structure.

### Encoding multi-locus data

Now imagine a different study that encodes complete haplotypes:

```{r, echo=FALSE}
input_counts |>
  filter(study_key == "study_02") |>
  kbl(format = "html", table.attr = "style='width:100%; white-space: nowrap;'") |>
  kable_styling(bootstrap_options = c("hover", "condensed"), 
                full_width = TRUE)
```

The first two rows show results for samples that amplified successfully over loci 72 to 76. Notice that the denominator of 65 is the same between these two rows. In general, the denominator *must be identical over all rows that share the same combination of genes and positions*. The same 65 samples could have produced either of these two variants, and so they are grouped together in the same denominator class.

The next two rows describe samples that only amplified at a subset of codons. The denominator gives the number of samples that amplified at each subset. This does not have to match the other subsets as it relates to a different combination of genes and positions.

Importantly, **there is no double counting in this table**. The `crt:72_73_74_75_76:CVIET` samples are not included in the `crt:72_73_74:CVI` numbers, even though technically they did amplify at all these loci. If you're ever unsure what to do, remember that **every sample is present in this table only once**.

For the same reason, it would be incorrect for this table to also list single-locus prevalences calculated from the full haplotypes. STAVE is smart enough to calculate single-locus prevalences from the full haplotype data, and so encoding the same information twice would result in double-counting.

### Encoding mixed calls

This third example has muti-locus haplotypes as well as mixed calls at some loci:

```{r, echo=FALSE}
input_counts |>
  filter(study_key == "study_03") |>
  kbl(format = "html", table.attr = "style='width:100%; white-space: nowrap;'") |>
  kable_styling(bootstrap_options = c("hover", "condensed"), 
                full_width = TRUE)
```

See the [variant string package](https://github.com/mrc-ide/variantstring) for further details of how to encode mixed calls, but in short, the `/` symbol specifies an un-phased mixed call, and `|` a phased mixed call.

All of these samples amplified successfully at positions 72 to 74, therefore the denominator of 100 is shared over all rows. The various combinations of mixed calls are each given a different row. Samples that are `crt:72_73_74:C/S_VI` are not also contained in the numbers for `crt:72_73_74:CVI` - as always, samples appear only once.

### How many loci is too many?

Consider this final example, in which phased information is presented over multiple genes:

```{r, echo=FALSE}
input_counts |>
  filter(study_key == "study_04") |>
  kbl(format = "html", table.attr = "style='width:100%; white-space: nowrap;'") |>
  kable_styling(bootstrap_options = c("hover", "condensed"), 
                full_width = TRUE)
```

This is a very long variant string. On the one hand, encoding the full haplotype retains all information. On the other hand, this notation starts to become quite cumbersome. More importantly, if we took this approach for a large number of samples then we would reach a combinatorial explosion in the number of possible observed variants and positions, as each distinct variant would require a new row.

A different approach might be to break this into different genes:

```{r, echo=FALSE}
input_counts |>
  filter(study_key == "study_05") |>
  kbl(format = "html", table.attr = "style='width:100%; white-space: nowrap;'") |>
  kable_styling(bootstrap_options = c("hover", "condensed"), 
                full_width = TRUE)
```

This would reduce the number of possible combinations, as new rows would only be needed to capture diversity in each gene. However, we have lost the ability to query the prevalence of haplotypes spanning multiple genes. For example, we may be interested in how frequently the *dhfr* and *dhps* mutations are found together, which is now lost to us because of how we have broken up the data.

Overall, there is a balance to strike between capturing everything and capturing enough. This balance will vary from one application to the next. In general, if your aim is to capture genome-wide haplotypes then STAVE is probably not the best solution, and you may be better sticking with other common formats like .vcf and .pmo. If your aim is to capture key drug resistance combinations in a lightweight and flexible notation, then STAVE may be a good choice.


---

The [next page](articles/calculating_prevalence.html) goes into the details of calculating prevalence from the encoded data.
