<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="STAVE">
<title>Encoding Genetic Data • STAVE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Encoding Genetic Data">
<meta property="og:description" content="STAVE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">STAVE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-how-it-works">How it works</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-how-it-works">
    <a class="dropdown-item" href="../articles/relational_structure.html">The relational data structure</a>
    <a class="dropdown-item" href="../articles/input_formats.html">Input formats</a>
    <a class="dropdown-item" href="../articles/variant_strings.html">Encoding genetic data</a>
    <a class="dropdown-item" href="../articles/calculating_prevalence.html">Calculating prevalence</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/reading_in_data.html">Reading in data and calculating prevalence</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mrc-ide/stave">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="counts_format_files/kePrint-0.0.1/kePrint.js"></script><link href="counts_format_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Encoding Genetic Data</h1>
            
      
      
      <div class="d-none name"><code>counts_format.Rmd</code></div>
    </div>

    
    
<p>One of the more nuanced aspects of using STAVE is learning how to
correctly format genetic data in the <em>Counts</em> table. STAVE is
designed with the flexibility to encode full haplotypes, preserving
information about linkage across multiple genes and multiple loci. But
what happens when not all loci amplify in some samples? Or when
published data only report locus-by-locus results, rather than full
haplotypes? These situations are entirely manageable once we understand
the underlying data structure.</p>
<div class="section level3">
<h3 id="encoding-single-locus-data">Encoding single-locus data<a class="anchor" aria-label="anchor" href="#encoding-single-locus-data"></a>
</h3>
<p>Imagine you come across a paper detailing <em>pfcrt</em> mutations at
codons 72 to 76 for a sample of 100 individuals. The CVIET haplotype,
commonly observed at these positions, is well-established as a marker of
chloroquine resistance. However, rather than presenting haplotype-level
data, the paper provides only site-specific information: the number of
individuals tested at each codon and the frequency of each observed
variant:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_key
</th>
<th style="text-align:left;">
survey_key
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72:C
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:73:V
</td>
<td style="text-align:right;">
95
</td>
<td style="text-align:right;">
95
</td>
</tr>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:74:I
</td>
<td style="text-align:right;">
31
</td>
<td style="text-align:right;">
98
</td>
</tr>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:75:E
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
92
</td>
</tr>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:76:T
</td>
<td style="text-align:right;">
45
</td>
<td style="text-align:right;">
89
</td>
</tr>
</tbody>
</table>
<p>Notice that the <code>total_num</code> is not consistent over loci.
This is very common, and is usually caused by some loci failing to
amplify in some samples.</p>
<p>This method of encoding genetic data is straightforward and easy to
interpret. For instance, the prevalence of glutamic acid (“E”) at codon
75 can be directly calculated as 20 out of 92 samples, or approximately
22%. However, this approach sacrifices information about full
haplotypes. Specifically, we cannot determine whether the 20 samples
with <code>crt:75:E</code> are the same individuals who also carry
<code>crt:76:T</code>. As a result, we are limited to estimating
single-locus prevalences and cannot infer multi-locus patterns or
linkage. While this encoding is valid, it is suboptimal for analyses
that rely on haplotype structure.</p>
</div>
<div class="section level3">
<h3 id="encoding-multi-locus-data">Encoding multi-locus data<a class="anchor" aria-label="anchor" href="#encoding-multi-locus-data"></a>
</h3>
<p>Now imagine a different study that encodes complete haplotypes:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_key
</th>
<th style="text-align:left;">
survey_key
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74_75_76:CVIET
</td>
<td style="text-align:right;">
23
</td>
<td style="text-align:right;">
65
</td>
</tr>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74_75_76:CVMET
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
65
</td>
</tr>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74:CVI
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:74_75_76:IET
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
3
</td>
</tr>
</tbody>
</table>
<p>The first two rows show results for samples that amplified
successfully over loci 72 to 76. Notice that the denominator of 65 is
the same between these two rows. In general, the denominator <em>must be
identical over all rows that share the same combination of genes and
positions</em>. The same 65 samples could have produced either of these
two variants, and so they are grouped together in the same denominator
class.</p>
<p>The next two rows describe samples that only amplified at a subset of
codons. The denominator gives the number of samples that amplified at
each subset. This does not have to match the other subsets as it relates
to a different combination of genes and positions.</p>
<p>Importantly, <strong>there is no double counting in this
table</strong>. The <code>crt:72_73_74_75_76:CVIET</code> samples are
not included in the <code>crt:72_73_74:CVI</code> numbers, even though
technically they did amplify at all these loci. If you’re ever unsure
what to do, remember that <strong>every sample is present in this table
only once</strong>.</p>
<p>For the same reason, it would be incorrect for this table to also
list single-locus prevalences calculated from the full haplotypes. STAVE
is smart enough to calculate single-locus prevalences from the full
haplotype data, and so encoding the same information twice would result
in double-counting.</p>
</div>
<div class="section level3">
<h3 id="encoding-mixed-calls">Encoding mixed calls<a class="anchor" aria-label="anchor" href="#encoding-mixed-calls"></a>
</h3>
<p>This third example has muti-locus haplotypes as well as mixed calls
at some loci:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_key
</th>
<th style="text-align:left;">
survey_key
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_03
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74:CVI
</td>
<td style="text-align:right;">
78
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_03
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74:C/S_VI
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_03
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74:C/S_V_I/M
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_03
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74:C|S_V_I|M
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
100
</td>
</tr>
</tbody>
</table>
<p>See the <a href="https://github.com/mrc-ide/variantstring" class="external-link">variant
string package</a> for further details of how to encode mixed calls, but
in short, the <code>/</code> symbol specifies an un-phased mixed call,
and <code>|</code> a phased mixed call.</p>
<p>All of these samples amplified successfully at positions 72 to 74,
therefore the denominator of 100 is shared over all rows. The various
combinations of mixed calls are each given a different row. Samples that
are <code>crt:72_73_74:C/S_VI</code> are not also contained in the
numbers for <code>crt:72_73_74:CVI</code> - as always, samples appear
only once.</p>
</div>
<div class="section level3">
<h3 id="how-many-loci-is-too-many">How many loci is too many?<a class="anchor" aria-label="anchor" href="#how-many-loci-is-too-many"></a>
</h3>
<p>Consider this final example, in which phased information is presented
over multiple genes:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_key
</th>
<th style="text-align:left;">
survey_key
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody><tr>
<td style="text-align:left;">
study_04
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74_75_76:CVIET;dhfr:51_59_108_164:NCSI;dhps:436_437_540_581_613:AGEGT
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
5
</td>
</tr></tbody>
</table>
<p>This is a very long variant string. On the one hand, encoding the
full haplotype retains all information. On the other hand, this notation
starts to become quite cumbersome. More importantly, if we took this
approach for a large number of samples then we would reach a
combinatorial explosion in the number of possible observed variants and
positions, as each distinct variant would require a new row.</p>
<p>A different approach might be to break this into different genes:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_key
</th>
<th style="text-align:left;">
survey_key
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_05
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
crt:72_73_74_75_76:CVIET
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
study_05
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
dhfr:51_59_108_164:NCSI
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
study_05
</td>
<td style="text-align:left;">
site_01
</td>
<td style="text-align:left;">
dhps:436_437_540_581_613:AGEGT
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
5
</td>
</tr>
</tbody>
</table>
<p>This would reduce the number of possible combinations, as new rows
would only be needed to capture diversity in each gene. However, we have
lost the ability to query the prevalence of haplotypes spanning multiple
genes. For example, we may be interested in how frequently the
<em>dhfr</em> and <em>dhps</em> mutations are found together, which is
now lost to us because of how we have broken up the data.</p>
<p>Overall, there is a balance to strike between capturing everything
and capturing enough. This balance will vary from one application to the
next. In general, if your aim is to capture genome-wide haplotypes then
STAVE is probably not the best solution, and you may be better sticking
with other common formats like .vcf and .pmo. If your aim is to capture
key drug resistance combinations in a lightweight and flexible notation,
then STAVE may be a good choice.</p>
<hr>
<p>The <a href="articles/calculating_prevalence.html">next page</a> goes
into the details of calculating prevalence from the encoded data.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
