<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="STAVE">
<title>Cautionary Notes • STAVE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Cautionary Notes">
<meta property="og:description" content="STAVE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">STAVE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-how-it-works">How it works</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-how-it-works">
    <a class="dropdown-item" href="../articles/howto_core_design_principles.html">Core Design Principles</a>
    <a class="dropdown-item" href="../articles/howto_studies_table.html">The Studies Table</a>
    <a class="dropdown-item" href="../articles/howto_surveys_table.html">The Surveys Table</a>
    <a class="dropdown-item" href="../articles/howto_counts_table.html">The Counts Table</a>
    <a class="dropdown-item" href="../articles/howto_encoding_data.html">How To Encode Genetic Data</a>
    <a class="dropdown-item" href="../articles/howto_calculating_prevalence.html">Calculating prevalence</a>
    <a class="dropdown-item" href="../articles/howto_cautionary_notes.html">Cautionary Notes</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/tutorial_reading_in_data.html">Reading in Data</a>
    <a class="dropdown-item" href="../articles/tutorial_calculating_prevalence.html">Calculating Prevalence</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mrc-ide/stave">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="howto_cautionary_notes_files/kePrint-0.0.1/kePrint.js"></script><link href="howto_cautionary_notes_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Cautionary Notes</h1>
            
      
      
      <div class="d-none name"><code>howto_cautionary_notes.Rmd</code></div>
    </div>

    
    
<p>Although STAVE rigorously defines the data structure, there are still
ways of obtaining misleading results depending on how the data are
entered. Even worse, given that STAVE encourages the use of exact
spatial coordinates, this structure has the potential to identify
individuals suffering from malaria, which has ethical implications. This
section highlights some common pitfalls and how to avoid them.</p>
<div class="section level2">
<h2 id="do-i-need-to-record-wild-type-samples">Do I need to record wild type samples?<a class="anchor" aria-label="anchor" href="#do-i-need-to-record-wild-type-samples"></a>
</h2>
<p>The table below encodes not only the mutant haplotypes of interest
but also how often the wild-type <em>pfcrt</em> haplotype
(<code>CVMNK</code>) was observed.</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_id
</th>
<th style="text-align:left;">
survey_id
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
study_01_site_01
</td>
<td style="text-align:left;">
crt:72-76:CVIET
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
study_01_site_01
</td>
<td style="text-align:left;">
crt:72-76:SVMNT
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
study_01_site_01
</td>
<td style="text-align:left;">
crt:72-76:CVMNK
</td>
<td style="text-align:right;">
70
</td>
<td style="text-align:right;">
100
</td>
</tr>
</tbody>
</table>
<p>Notice that the sum of the numerators across all variants equals the
denominator. This tells us that the entire space of possible haplotypes
is covered. Because the data are complete, we can query the prevalence
of any allele at any locus and obtain a correct result. For example,
here is the prevalence of the wild-type allele (<code>C</code>) at codon
72:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>s<span class="sc">$</span><span class="fu">get_prevalence</span>(<span class="at">target_variant =</span> <span class="st">"crt:72:C"</span>) <span class="sc">|&gt;</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>  ... <span class="co"># (only show some columns)</span></span></code></pre></div>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
survey_id
</th>
<th style="text-align:right;">
numerator
</th>
<th style="text-align:right;">
denominator
</th>
<th style="text-align:right;">
prevalence
</th>
<th style="text-align:right;">
prevalence_lower
</th>
<th style="text-align:right;">
prevalence_upper
</th>
</tr></thead>
<tbody><tr>
<td style="text-align:left;">
study_01_site_01
</td>
<td style="text-align:right;">
90
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
90
</td>
<td style="text-align:right;">
82.37774
</td>
<td style="text-align:right;">
95.09953
</td>
</tr></tbody>
</table>
<p>This is correct - the <code>C</code> allele was seen 20 times in the
<code>CVIET</code> mutant haplotype and 70 times in the WT
<code>CVMNK</code> haplotype. Therefore the prevalence is (20 + 70) /
100 = 90%.</p>
<p>However, many studies do not report how many samples were wild-type.
Instead, they only report “interesting” haplotypes — typically those
carrying resistance-associated mutations. In the dataset below, the only
difference is that the WT observations have been omitted:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_id
</th>
<th style="text-align:left;">
survey_id
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
study_02_site_01
</td>
<td style="text-align:left;">
crt:72-76:CVIET
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
study_02_site_01
</td>
<td style="text-align:left;">
crt:72-76:SVMNT
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
100
</td>
</tr>
</tbody>
</table>
<p>What happens if we now attempt to estimate the prevalence of 72
C?</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>s<span class="sc">$</span><span class="fu">get_prevalence</span>(<span class="at">target_variant =</span> <span class="st">"crt:72:C"</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  ... <span class="co"># (only show some columns)</span></span></code></pre></div>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:right;">
numerator
</th>
<th style="text-align:right;">
denominator
</th>
<th style="text-align:right;">
prevalence
</th>
<th style="text-align:right;">
prevalence_lower
</th>
<th style="text-align:right;">
prevalence_upper
</th>
</tr></thead>
<tbody><tr>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
12.66556
</td>
<td style="text-align:right;">
29.18427
</td>
</tr></tbody>
</table>
<p>The result is clearly wrong: the estimate of 20% is far too low. This
happens because, once the WT haplotype is omitted, the only remaining
records of the <code>C</code> allele at codon 72 come from samples that
were mutant somewhere else in the haplotype. All the true WT samples are
invisible to STAVE, so they cannot be counted.</p>
<p>This naturally raises the question: <em>Should we always record WT
haplotypes so that numerators sum to the denominator?</em></p>
<p>In an ideal world, yes. If the complete distribution of haplotypes is
known, including all WT observations, many downstream queries become
trivial and robust.</p>
<p>In practice, this is often impossible because:</p>
<ul>
<li>the authors focused on certain mutations, meaning we cannot assume
all unreported haplotypes are WT,</li>
<li>the reference genome used in analysis is not stated,</li>
<li>the data have already been summarised before reaching you.</li>
</ul>
<p>If WT samples were not recorded, STAVE cannot invent them. The
resulting prevalence estimates will therefore be biased toward the
mutants, and uncertainty intervals will reflect only the partial
data.</p>
<div class="section level3">
<h3 id="practical-implications">Practical implications<a class="anchor" aria-label="anchor" href="#practical-implications"></a>
</h3>
<ul>
<li>If the WT count is available, include it — doing so prevents many
errors.</li>
<li>If it is not available, be cautious when querying allele-level
prevalence, especially for positions that are wild-type in the reference
haplotype (e.g. <em>mdr1</em> N86).</li>
<li>When only mutant haplotypes are reported, the safest queries are
those that target exactly the mutations that were reported, not their
complements.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="restricted-or-identifying-data">Restricted or identifying data<a class="anchor" aria-label="anchor" href="#restricted-or-identifying-data"></a>
</h2>
<p>STAVE is designed to handle both public and non-public datasets. The
access_level field allows users to mark each study as public,
restricted, or private, and STAVE itself places no limitations on
loading or analysing restricted material. This design choice allows
individual laboratories and surveillance teams to use STAVE with their
own internal datasets, even when those datasets cannot be shared
publicly.</p>
<p>However, this also means that <strong>users are entirely responsible
for ensuring that STAVE objects containing restricted or private data do
not enter the public domain</strong>. STAVE provides the structure, but
it cannot enforce data-governance rules; those must be respected by the
analysts who use it.</p>
<p>A second consideration relates to identifiability. Aggregate genetic
data are often assumed to be non-identifying because they represent
groups rather than individuals. However, STAVE encourages
high-resolution spatial recording of sampling locations, and at
sufficiently fine granularity this assumption may break down. For
example, if a single individual presents to a precisely geo-referenced
health facility on a specific day, those data could be identifying even
in aggregated form.</p>
<p>For this reason, <strong>users must ensure that the spatial and
temporal resolution of their data is ethically and legally
appropriate</strong>. This may require intentionally reducing precision
— for example, by jittering, rounding, or otherwise pseudonymising
coordinates before constructing a STAVE object. If such steps are taken,
they should be clearly documented in the <code>location_notes</code>
(and, if relevant, <code>time_notes</code>) fields so that any
downstream user understands the provenance and limitations of the
spatial information.</p>
<hr>
<p>If you have read all the documentation up to this point, then bless
you! You should have a good understanding of how STAVE works, and are
ready to jump to the <a href="articles/installation.html">Installation</a> and <a href="articles/reading_in_data.html">Tutorials</a> sections to start
putting some of these ideas into practice.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
