<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="STAVE">
<title>Calculating Prevalence • STAVE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Calculating Prevalence">
<meta property="og:description" content="STAVE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">STAVE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-how-it-works">How it works</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-how-it-works">
    <a class="dropdown-item" href="../articles/howto_core_design_principles.html">Core Design Principles</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>The Relational Data Structure</h6>
    <a class="dropdown-item" href="../articles/howto_studies_table.html">The Studies Table</a>
    <a class="dropdown-item" href="../articles/howto_surveys_table.html">The Surveys Table</a>
    <a class="dropdown-item" href="../articles/howto_counts_table.html">The Counts Table</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/howto_encoding_data.html">How To Encode Genetic Data</a>
    <a class="dropdown-item" href="../articles/howto_calculating_prevalence.html">Calculating prevalence</a>
    <a class="dropdown-item" href="../articles/howto_cautionary_notes.html">Cautionary Notes</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/tutorial_reading_in_data.html">Reading in Data</a>
    <a class="dropdown-item" href="../articles/tutorial_calculating_prevalence.html">Calculating Prevalence</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mrc-ide/stave">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="howto_calculating_prevalence_files/kePrint-0.0.1/kePrint.js"></script><link href="howto_calculating_prevalence_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Calculating Prevalence</h1>
            
      
      
      <div class="d-none name"><code>howto_calculating_prevalence.Rmd</code></div>
    </div>

    
    
<p>In previous pages we learned how to encode genetic data in STAVE. Now
we turn to the problem of calculating prevalence from the encoded
data.</p>
<p>In simple terms, prevalence is calculated as the number of times a
variant is observed divided by the number of times it <em>could have
been</em> observed. However, nuances in prevalence calculation warrant
further discussion — particularly in cases of mixed (heterozygous)
calls, where it may only be possible to estimate a prevalence range
rather than an exact value.</p>
<div class="section level3">
<h3 id="matching-variant-strings">Matching variant strings<a class="anchor" aria-label="anchor" href="#matching-variant-strings"></a>
</h3>
<p>Imagine the following data are loaded:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_id
</th>
<th style="text-align:left;">
survey_id
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
study_01_survey_01
</td>
<td style="text-align:left;">
crt:76:T
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
20
</td>
</tr>
<tr>
<td style="text-align:left;">
study_01
</td>
<td style="text-align:left;">
study_01_survey_01
</td>
<td style="text-align:left;">
crt:75_76:ET
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
5
</td>
</tr>
</tbody>
</table>
<p>Prevalence calculation in STAVE starts by specifying the variant of
interest. For example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s</span><span class="op">$</span><span class="fu">get_prevalence</span><span class="op">(</span>target_variant <span class="op">=</span> <span class="st">"crt:76:T"</span><span class="op">)</span></span></code></pre></div>
<p>This is then compared against all data loaded into the current
object.</p>
<p>For example, <code>crt:76:T</code> is clearly a direct match to the
first row, and so these values should be added to the numerator and
denominator. It is also a match to the second row, which contains
<code>crt:76:T</code> as a subset, meaning these values should also be
added to the numerator and denominator. This gives prevalence = (4 + 1)
/ (20 + 5) = 20%.</p>
<p>Now imagine we want to know the prevalence of <code>crt:76:E</code>.
This is only a match to the second row in terms of the variant, however,
it is a match to both rows in terms of position. In other words, both
rows <em>could</em> have produced allele <code>E</code> at position 76.
Hence, we obtain prevalence = 1 / (20 + 5) = 4%.</p>
<p>Finally, imagine we want to know the prevalence of the
<code>crt:75_76:ET</code> haplotype. This is only a match to the second
row in both numerator and denominator. Hence, we obtain prevalence = 1 /
5 = 20%.</p>
</div>
<div class="section level3">
<h3 id="dealing-with-unphased-mixed-calls">Dealing with unphased mixed calls<a class="anchor" aria-label="anchor" href="#dealing-with-unphased-mixed-calls"></a>
</h3>
<p>Now let’s look at an example where there are unphased mixed
calls:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_id
</th>
<th style="text-align:left;">
survey_id
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
study_02_survey_01
</td>
<td style="text-align:left;">
crt:72-76:CVIET
</td>
<td style="text-align:right;">
50
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
study_02_survey_01
</td>
<td style="text-align:left;">
crt:72-76:CVMNK
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_02
</td>
<td style="text-align:left;">
study_02_survey_01
</td>
<td style="text-align:left;">
crt:72-76:CV_I/M_E/N_T/K
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
100
</td>
</tr>
</tbody>
</table>
<p>Here we have two common <em>pfcrt</em> haplotypes, CVIET and CVMNK,
as well as some mixed calls.</p>
<p>Imagine we want to know the prevalence of <code>crt:72-74:CVI</code>.
This is a match to the first row, and interestingly it is also an
<em>unambiguous</em> match to the third row. The third variant
<code>crt:72-76:CV_I/M_E/N_T/K</code> contains multiple unphased
heterozygous sites, meaning we cannot work out the exact component
haplotypes that make up this mixture. However, looking solely at
positions 72 to 74, we know for certain that the haplotypes CVI and CVM
are present. Therefore we can be certain that this matches our target.
We obtain prevalence = (50 + 10) / 100 = 60%.</p>
<p>Now imagine we want to know the prevalence of the full
<code>crt:72-76:CVMNK</code> haplotype. This matches the second row, but
because there are multiple unphased heterozygous sites in the third row,
it is only an <em>ambiguous</em> match to this variant. Our target
<em>may</em> be present in these samples, but equally it may not. This
presents a problem - if we include this in the numerator then we risk
overestimating the prevalence, but if we exclude it then we risk
underestimating the prevalence. Faced with this dilemma, the approach
taken in STAVE is to simply report the <strong>minimum and maximum
possible prevalence</strong> implied by the data. This would give
<code>min_prevalence</code> = 40 / 100 = 40%,
<code>max_prevalence</code> = (40 + 10) / 100 = 50%. It is then up to
the user to decide what to do with this information. Valid approaches
include taking the min, the max, the midpoint, or doing more advanced
statistical imputation based on per-locus prevalence. Whatever choice is
made, it should be clearly documented.</p>
</div>
<div class="section level3">
<h3 id="dealing-with-phased-mixed-calls">Dealing with phased mixed calls<a class="anchor" aria-label="anchor" href="#dealing-with-phased-mixed-calls"></a>
</h3>
<p>Finally, imagine the same data as before, but now the mixed haplotype
is encoded as phased information:</p>
<table style="width:100%; white-space: nowrap; margin-left: auto; margin-right: auto;" class="table table table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
study_id
</th>
<th style="text-align:left;">
survey_id
</th>
<th style="text-align:left;">
variant_string
</th>
<th style="text-align:right;">
variant_num
</th>
<th style="text-align:right;">
total_num
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
study_03
</td>
<td style="text-align:left;">
study_03_survey_01
</td>
<td style="text-align:left;">
crt:72-76:CVIET
</td>
<td style="text-align:right;">
50
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_03
</td>
<td style="text-align:left;">
study_03_survey_01
</td>
<td style="text-align:left;">
crt:72-76:CVMNK
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
study_03
</td>
<td style="text-align:left;">
study_03_survey_01
</td>
<td style="text-align:left;">
crt:72-76:CV_I|M_E|N_T|K
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
100
</td>
</tr>
</tbody>
</table>
<p>The mixtures are now clearly specified as a combination of the CVIET
and CVMNK haplotypes. If our target is <code>crt:72-76:CVMNK</code> as
before, this is now an unambiguous match against the second and third
rows, giving prevalence = (40 + 10) / 100 = 50%. This is a strong
argument for including phased data wherever possible, although in
reality it can be challenging to reliably phase genomes.</p>
<p>More complex situations, such as partial phasing or more than two
alleles at a locus, are allowed - see the <a href="https://github.com/mrc-ide/variantstring" class="external-link">variantstring package
documentation</a> for details.</p>
<hr>
<p>The <a href="articles/cautionary_notes.html">next page</a> goes on to
explain some of the ways that we must be careful when entering or
interpreting data.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
