<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="STAVE">
<title>Encoding Variants as Strings • STAVE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Encoding Variants as Strings">
<meta property="og:description" content="STAVE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">STAVE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-how-it-works">How it works</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-how-it-works">
    <a class="dropdown-item" href="../articles/relational_structure.html">The relational data structure</a>
    <a class="dropdown-item" href="../articles/input_formats.html">Input data formats</a>
    <a class="dropdown-item" href="../articles/variant_strings.html">Encoding variants as strings</a>
    <a class="dropdown-item" href="../articles/calculating_prevalence.html">Calculating prevalence</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../TODO">Reading in data</a>
    <a class="dropdown-item" href="../TODO">Calculating prevalence and other outputs</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mrc-ide/stave">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Encoding Variants as Strings</h1>
            
      
      
      <div class="d-none name"><code>variant_strings.Rmd</code></div>
    </div>

    
    
<p>Storing genetic information presents unique challenges, particularly
when dealing with varying levels of detail reported by different
sources. For instance, some academic studies may only provide aggregate
counts of specific amino acids observed at a single locus, whereas
others might supply raw sequence data at the individual level, allowing
us to perform our own aggregation. To address these differences, we need
a data format that is both versatile and robust.</p>
<p>The solution in STAVE is a character string format that adheres to
strict rules. While this approach may not be the most efficient for data
compression, this trade-off is less significant when working with
aggregate data, where the number of distinct observations is
limited.</p>
<p>This page explains the variant string format in detail, providing
clear examples of both compatible and incompatible data types.</p>
<div class="section level3">
<h3 id="gene-name-locus-and-amino-acid-are-separated-by">1. Gene name, locus, and amino acid are separated by ;<a class="anchor" aria-label="anchor" href="#gene-name-locus-and-amino-acid-are-separated-by"></a>
</h3>
<p>These are the three key elements. For example, the string
<em>pfk13;580;Y</em> specifies that in the <em>pfk13</em> gene (the
kelch 13 gene, implicated in Artemisinin resistance) at the 580th codon
position, we observed the amino acid Y. Amino acids must follow the <a href="https://megasoftware.net/web_help_11/IUPAC_Single_Letter_Codes.htm" class="external-link">IUPAC
single-letter format</a>, in this case Y corresponds to Tyrosine.</p>
<p>Another common notation would call this the <em>C580Y</em> variant.
We refrain from this notation because:</p>
<ol style="list-style-type: decimal">
<li>This notation does not scale well to multiple loci.</li>
<li>Sources do not always report the wild type (Cysteine in this case),
forcing us to do extra work tracking down the reference genome.</li>
<li>Strictly, we cannot assume that all non-numerator observations are
wild type. The denominator may include all other variants, in which case
recording these as wild type would be a corruption of the original
data.</li>
</ol>
<p>Gene names must consist solely of lowercase English letters and
digits (0-9), without hyphens or other special characters. Otherwise,
they do not follow a strict convention. We recognise that this creates
the potential for ambiguity, for example if two sources use slightly
different names for the same gene, such as <em>pfk13</em> and
<em>k13</em>. However, this can be avoided by manually listing all gene
names in the existing data using the <code>get_variants()</code>
function before appending new data.</p>
</div>
<div class="section level3">
<h3 id="loci-are-separated-by-_">2. Loci are separated by _<a class="anchor" aria-label="anchor" href="#loci-are-separated-by-_"></a>
</h3>
<p>In some cases, sources may report variants over multiple loci. A good
example is the <em>pfcrt</em> gene, implicated in resistance to
Amodiaquine among other antimalarials, in which different haplotypes
confer different degrees of resistance. Here, the variant string
<em>pfcrt;72_73_74_75_76;C_V_I_E_T</em> specifies that at codon 72 we
observed a C (Cysteine), at codon 73 we observed a V (Valine), etc.</p>
<p>In a more concise shorthand notation, the underscores between amino
acids can be omitted, resulting in <em>pfcrt;72_73_74_75_76;CVIET</em>.
However, it is essential to retain the underscores between codon
positions to prevent ambiguity in the numbers.</p>
<p>When a data source provides a full haplotype, you should
<strong>always store the haplotype as a whole</strong>. Do not manually
break this down into individual loci, for example one entry for
<em>pfcrt;72;C</em>, then a separate entry for <em>pfcrt;73;V</em> etc.
When calculating prevalence, STAVE is smart enough to know that the
<em>pfcrt;72;C</em> variant is contained within the
<em>pfcrt;72_73_74_75_76;CVIET</em> haplotype. Even more importantly,
<strong>do not store both the haplotype and the single locus
information</strong>, as this will lead to mistakes in prevalence
calculation due to double-counting. A more detailed explanation of this
issue, along with guidance on handling very long haplotypes derived from
raw sequence data, is provided on the <a href="articles/calculating_prevalence.html">next page</a>.</p>
</div>
<div class="section level3">
<h3 id="genes-are-separated-by">3. Genes are separated by :<a class="anchor" aria-label="anchor" href="#genes-are-separated-by"></a>
</h3>
<p>We may want to capture information about mutations observed together
across multiple genes. A key example is the antimalarial combination
therapy sulfadoxine-pyrimethamine (SP), where resistance arises from
mutations in two genes:</p>
<ul>
<li>Mutations in the <em>pfdhfr</em> gene reduce the parasite’s
susceptibility to pyrimethamine.</li>
<li>Mutations in the <em>pfdhps</em> gene reduce susceptibility to
sulfadoxine.</li>
</ul>
<p>This can be represented in the variant string format by listing
multiple genes separated by a colon. For example:
<em>pfdhfr;51;I:pfdhps;437;G</em>.</p>
<p>The order in which genes are listed does not matter, as STAVE will
automatically sort them before calculating prevalence. Additionally,
there is no restriction on the number of genes that can be included in a
single variant string.</p>
</div>
<div class="section level3">
<h3 id="unphased-mixed-calls-are-indicated-by">4. Unphased mixed calls are indicated by /<a class="anchor" aria-label="anchor" href="#unphased-mixed-calls-are-indicated-by"></a>
</h3>
<p>Polyclonal infections are common in malaria and can result in mixed,
or “heterozygous,” calls at the DNA level. In some cases, these mixed
calls also manifest at the amino acid level. When this occurs, all
observed amino acids should be listed, separated by a forward slash.</p>
<p>For example, the string <em>pfmdr1;76;N/Y</em> indicates that in the
<em>pfmdr1</em> gene at codon position 76, both N (Asparagine) and Y
(Tyrosine) were observed.</p>
<p>Any number of alleles can be listed, such as
<em>pfmdr1;76;N/Y/F</em>, indicating three amino acids — N, Y, and F
(Phenylalanine) — were detected at position 76. These mixed calls can
also be incorporated into multi-locus notation. For instance,
<em>pfmdr1;76_184;N/Y_Y</em> specifies a heterozygous call at position
76 (N/Y) and a homozygous call at position 184 (Y).</p>
</div>
<div class="section level3">
<h3 id="phased-mixed-calls-are-indicated-by">5. Phased mixed calls are indicated by |<a class="anchor" aria-label="anchor" href="#phased-mixed-calls-are-indicated-by"></a>
</h3>
<p>The forward slash notation assumes that variants are unphased,
meaning we do not know which variants go together. For example, in the
string <em>pfmdr1;76_184;N/Y_Y/F</em> we do not know if this is caused
by the haplotype combination <em>N_Y</em> + <em>Y_F</em>, or the
combination <em>N_F</em> + <em>Y_Y</em>. There could also be more than
two genotypes present in the sample, in which case more combinations are
possible.</p>
<p>In some cases, we may have phased information, meaning we know which
variants go together. These can be indicated using the vertical bar
symbol in place of the forward slash. When using this notation, the
first letter listed at a locus always goes with the first letter at
another locus, etc. For example, <em>pfmdr1;76_184;N|Y_Y|F</em>
indicates the haplotype combination <em>N_Y</em> + <em>Y_F</em>.</p>
<p>Any number of alleles can be listed, but the same number of alleles
must be present at all heterozygous loci. For example, if we have
observed three distinct amino acids at position 76 and two distinct
amino acids at position 184, then we must list three amino acids at
locus 184 even if this means repeating the same letter. Otherwise it
would not be possible to link up which amino acids go together.</p>
<p>There are some limitations to this format. For example, phased and
unphased regions are not currently supported in the same variant.
Similarly, partial phasing is not supported, by which we mean samples
with more than two genomes where a only subset of genomes are phased
together. As a rule of thumb, remember that <strong>the / and | symbols
cannot be used together</strong>. While we recognise this as a
limitation of the format, these are relatively rare edge cases. They
will also not tend to impact prevalence calculation unless we are
interested in the prevalence of a long haplotype. In both cases above,
data can still be included in STAVE format with some information loss,
for example by treating all loci as unphased.</p>
</div>
<div class="section level3">
<h3 id="what-cannot-be-encoded">What cannot be encoded?<a class="anchor" aria-label="anchor" href="#what-cannot-be-encoded"></a>
</h3>
<p>As with any data format, there are inherent trade-offs in the variant
string encoding. The following information <strong>cannot</strong> be
encoded:</p>
<ul>
<li>Copy number variation</li>
<li>More complex genomic rearrangements, including insertions and
deletions</li>
<li>Combinations of phased and unphased regions and partial phasing (see
above)</li>
<li>Within-sample allele frequencies</li>
<li>Quality scores</li>
</ul>
<p>Keep in mind that STAVE is designed to facilitate data extraction
from diverse sources for the purposes of prevalence calculation. It is
not intended to replace individual-level data or raw sequence data,
which remain essential for other types of analyses.</p>
<p>The <a href="articles/calculating_prevalence.html">next page</a>
explains how prevalence is calculated from this encoding.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
